### PR TITLE
Feature/zan test filter

### PR DESCRIPTION
**Pull request by DoraALin**
pr for #64
1. zan test message filter in nsqd.
2. introduce allow-zan-test-skip option in nsqd to enable zan-test message filter after topic upgrade & creation
3. introduce  enable-zan-test-skip option in nsqadmin to enable/disable zan test skip button in channel page.

### CODE REVIEW

File: nsqd/topic.go [hunk 1]
```diff
@@ -646,7 +650,14 @@ func (t *Topic) SetDynamicInfo(dynamicConf TopicDynamicConf, idGen MsgIDGenerato
 		ext := dynamicConf.Ext
 		ch.SetExt(ext)
 		if ext && t.option.AllowZanTestSkip {
-			ch.SkipZanTest()
+			switch ch.ZanTestSkipped() {
+			case ZanTestSkip :
```
Comment (absolute8511) [inline nsqd/topic.go]: Do we need set zan test skip state here?

File: nsqd/channel.go [hunk 2]
```diff
@@ -181,6 +186,11 @@ func NewChannel(topicName string, part int, channelName string, chEnd BackendQue
 		peekedMsgs:             make([]Message, MaxWaitingDelayed),
 		Ext:                    ext,
 	}
+
+	if opt.AllowZanTestSkip && c.IsExt() {
```
Comment (absolute8511) [inline nsqd/channel.go]: Do we need set zan skip  state here?

File: nsqd/channel.go [hunk 3]
```diff
@@ -1852,8 +1891,11 @@ LOOP:
 			}
 		}
 
+		//check if topic is ext and message has zan test header
+		zanTestSkip := c.shouldSkipZanTest(msg)
```
Comment (absolute8511) [inline nsqd/channel.go]: should avoid `c.shouldSkipZanTest(msg)` if channel is skipped

File: nsqadmin/http.go [hunk 4]
```diff
@@ -1319,6 +1321,32 @@ func (s *httpServer) topicChannelAction(req *http.Request, topicName string, cha
...
 			s.notifyAdminActionWithUser("unskip_channel", topicName, channelName, "", req)
 		}
+	case "skipZanTest":
+		if channelName != "" {
+			err = s.ci.SkipZanTest(topicName, channelName,
+				s.ctx.nsqadmin.opts.NSQLookupdHTTPAddressesDC,
+				s.ctx.nsqadmin.opts.NSQDHTTPAddresses)
+
+			if err == nil {
+				err = s.ci.EmptyChannel(topicName, channelName,
```
Comment (absolute8511) [inline nsqadmin/http.go]: should not empty here

Comment (DoraALin) [reply nsqadmin/http.go]: empty channel removed

**Pull request merged**